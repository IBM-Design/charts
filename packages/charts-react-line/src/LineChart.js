import React, { PropTypes } from 'react';
import * as shape from 'd3-shape';
import Chart from '@ibm-design/charts-react-chart';
import Colors from 'ibm-design-colors/ibm-colors';
import Line from './Line';
import Legend from './Legend';

export default class LineChart extends React.PureComponent {
  static propTypes = {
    // eslint-disable-next-line react/forbid-prop-types
    lines: PropTypes.array.isRequired,
    // eslint-disable-next-line react/forbid-prop-types
    legend: PropTypes.shape({
      title: PropTypes.string,
      labels: PropTypes.array,
      width: PropTypes.number,
    }),
    width: PropTypes.number,
    height: PropTypes.number,
    margin: PropTypes.shape({
      top: PropTypes.number,
      right: PropTypes.number,
      bottom: PropTypes.number,
      left: PropTypes.number,
    }),
    x: PropTypes.func,
    y: PropTypes.func,
    scaleX: PropTypes.func,
    scaleY: PropTypes.func,
    // eslint-disable-next-line react/forbid-prop-types
    domainX: PropTypes.array,
    // eslint-disable-next-line react/forbid-prop-types
    domainY: PropTypes.array,
    // eslint-disable-next-line react/forbid-prop-types
    strokes: PropTypes.array,
  }

  static contextTypes = {
    chart: PropTypes.any,
  }

  static defaultProps = {
    lines: [],
    width: 960,
    height: 500,
    legend: {
      labels: [],
      width: 200,
    },
    margin: {
      top: 20,
      right: 20,
      bottom: 30,
      left: 50,
    },
    strokes: [
      Colors.blue['40'],
      Colors.aqua['20'],
      Colors.green['30'],
      Colors.lime['20'],
      Colors.gold['20'],
      Colors.peach['30'],
      Colors.magenta['40'],
      Colors.indigo['40'],
    ],
  }

  componentWillMount() {
    const {
      width,
      height,
      margin,
      x,
      y,
      scaleX,
      scaleY,
      domainX,
      domainY,
    } = this.props;

    this.width = width - margin.left - margin.right;
    this.height = height - margin.top - margin.bottom;

    this.x = scaleX()
      .rangeRound([0, this.width])
      .domain(domainX);

    this.y = scaleY()
      .rangeRound([this.height, 0])
      .domain(domainY);

    this.line = shape.line()
      .x((d) => this.x(x(d)))
      .y((d) => this.y(y(d)));
  }

  render() {
    const { line, props } = this;
    const { grid, height, legend, margin, strokes, width } = props;
    const isLegendVisible = !!legend.labels.length;
    const legendWidth = isLegendVisible ? (legend.width || 200) : 0;
    const lines = this.props.lines.map((data, index) =>
      <Line
        line = {line}
        data = {data}
        key={index}
        stroke={strokes[index % strokes.length]}
      />
    );

    return (
      <div style={{width}}>
        <style dangerouslySetInnerHTML={{__html: `
          @keyframes dash {
            to { stroke-dashoffset: 0; }
          }
        `}}
        />
        {isLegendVisible &&
          <Legend
            labels={legend.labels}
            width={legendWidth}
            title={legend.title}
          />
        }
        <Chart
          height={height}
          margin={margin}
          width={width - legendWidth}
          x={grid[0]}
          y={grid[1]}>
          {lines}
        </Chart>
      </div>
    );
  }
}
