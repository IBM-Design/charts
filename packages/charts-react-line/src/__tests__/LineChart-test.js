import React from 'react';
import { extent } from 'd3-array';
import { timeParse } from 'd3-time-format';
import { scaleLinear, scaleTime } from 'd3-scale';
import renderer from 'react-test-renderer';
import LineChart from '../LineChart';

const parseTime = timeParse('%d-%b-%y');
const data = [
  {
    close: 422.40,
    date: parseTime('6-Jan-12'),
  },
];

describe('LineChart Component', () => {
  it('should render', () => {
    const tree = renderer.create(
      <LineChart
        lines={[data]}
        width={960}
        height={500}
        margin={{
          top: 20,
          right: 20,
          bottom: 30,
          left: 50,
        }}
        scaleX={scaleTime}
        scaleY={scaleLinear}
        domainX={extent(data, (d) => d.date)}
        domainY={extent(data, (d) => d.close)}
        x={(d) => d.date}
        y={(d) => d.close}
        grid={[
          {
            text: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'June', 'July', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
          },
          {
            min: 0,
            max: 500,
            tickCount: 5,
          },
        ]}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
