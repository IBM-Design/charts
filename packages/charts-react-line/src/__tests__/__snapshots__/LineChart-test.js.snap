exports[`LineChart Component should render 1`] = `
<div
  style={
    Object {
      "width": 960,
    }
  }>
  <svg
    className={undefined}
    height={500}
    width={960}>
    <g
<<<<<<< HEAD
      className="axis axis--x"
      fill="none"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      transform="">
      <g
        className="tick"
        transform="translate(0, 450)">
        <path
          d="M0,0 L890,0"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.32em"
          fill="#000"
          x={-9}
          y={0.5}>
          0
        </text>
      </g>
      <g
        className="tick"
        transform="translate(0, 360)">
        <path
          d="M0,0 L890,0"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.32em"
          fill="#000"
          x={-9}
          y={0.5}>
          100
        </text>
      </g>
      <g
        className="tick"
        transform="translate(0, 270)">
        <path
          d="M0,0 L890,0"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.32em"
          fill="#000"
          x={-9}
          y={0.5}>
          200
        </text>
      </g>
      <g
        className="tick"
        transform="translate(0, 180)">
        <path
          d="M0,0 L890,0"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.32em"
          fill="#000"
          x={-9}
          y={0.5}>
          300
        </text>
      </g>
      <g
        className="tick"
        transform="translate(0, 90)">
        <path
          d="M0,0 L890,0"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.32em"
          fill="#000"
          x={-9}
          y={0.5}>
          400
        </text>
      </g>
      <g
        className="tick"
        transform="translate(0, 0)">
        <path
          d="M0,0 L890,0"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.32em"
          fill="#000"
          x={-9}
          y={0.5}>
          500
        </text>
      </g>
    </g>
    <g
      className="axis axis--x"
      fill="none"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      transform="translate(0,450)">
      <g
        className="tick"
        transform="translate(0, 0)">
        <path
          d="M0,0 L0,-450"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.71em"
          fill="#000"
          x={0.5}
          y={9}>
          Jan
        </text>
      </g>
      <g
        className="tick"
        transform="translate(80.9090909090909, 0)">
        <path
          d="M0,0 L0,-450"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.71em"
          fill="#000"
          x={0.5}
          y={9}>
          Feb
        </text>
      </g>
      <g
        className="tick"
        transform="translate(161.8181818181818, 0)">
        <path
          d="M0,0 L0,-450"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.71em"
          fill="#000"
          x={0.5}
          y={9}>
          Mar
        </text>
      </g>
      <g
        className="tick"
        transform="translate(242.72727272727272, 0)">
        <path
          d="M0,0 L0,-450"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.71em"
          fill="#000"
          x={0.5}
          y={9}>
          Apr
        </text>
      </g>
      <g
        className="tick"
        transform="translate(323.6363636363636, 0)">
        <path
          d="M0,0 L0,-450"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.71em"
          fill="#000"
          x={0.5}
          y={9}>
          May
        </text>
      </g>
      <g
        className="tick"
        transform="translate(404.5454545454545, 0)">
        <path
          d="M0,0 L0,-450"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.71em"
          fill="#000"
          x={0.5}
          y={9}>
          June
        </text>
      </g>
      <g
        className="tick"
        transform="translate(485.45454545454544, 0)">
        <path
          d="M0,0 L0,-450"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.71em"
          fill="#000"
          x={0.5}
          y={9}>
          July
        </text>
      </g>
      <g
        className="tick"
        transform="translate(566.3636363636364, 0)">
        <path
          d="M0,0 L0,-450"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.71em"
          fill="#000"
          x={0.5}
          y={9}>
          Aug
        </text>
      </g>
      <g
        className="tick"
        transform="translate(647.2727272727273, 0)">
        <path
          d="M0,0 L0,-450"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.71em"
          fill="#000"
          x={0.5}
          y={9}>
          Sep
        </text>
      </g>
      <g
        className="tick"
        transform="translate(728.1818181818181, 0)">
        <path
          d="M0,0 L0,-450"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.71em"
          fill="#000"
          x={0.5}
          y={9}>
          Oct
        </text>
      </g>
      <g
        className="tick"
        transform="translate(809.090909090909, 0)">
        <path
          d="M0,0 L0,-450"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.71em"
          fill="#000"
          x={0.5}
          y={9}>
          Nov
        </text>
      </g>
      <g
        className="tick"
        transform="translate(890, 0)">
        <path
          d="M0,0 L0,-450"
          fill="none"
          stroke="#CDDBF1"
          strokeWidth="1" />
        <text
          dy="0.71em"
          fill="#000"
          x={0.5}
          y={9}>
          Dec
        </text>
      </g>
    </g>
    <path
      d="M0,450Z"
      style={
        Object {
          "fill": "none",
          "stroke": "#5392ff",
          "strokeWidth": 1.5,
        }
      } />
  </g>
</svg>
=======
      transform="translate(50,20)">
      <g
        className="axis axis--y"
        fill="none"
        fontFamily="sans-serif"
        fontSize="10"
        textAnchor="end">
        <path
          className="domain"
          d="M-6,450H0V0.5H-6"
          stroke="#000"
          style={
            Object {
              "fill": "none",
              "shapeRendering": "crispEdges",
              "strokeWidth": 1,
            }
          } />
        <g
          className="tick"
          transform="translate(0,90)">
          <line
            stroke="#000"
            x2="-6"
            y1="0.5"
            y2="0.5" />
          <text
            dy="0.32em"
            fill="#000"
            x="-9"
            y="0.5">
            Label
          </text>
        </g>
        <g
          className="tick"
          transform="translate(0,180)">
          <line
            stroke="#000"
            x2="-6"
            y1="0.5"
            y2="0.5" />
          <text
            dy="0.32em"
            fill="#000"
            x="-9"
            y="0.5">
            Label
          </text>
        </g>
        <g
          className="tick"
          transform="translate(0,270)">
          <line
            stroke="#000"
            x2="-6"
            y1="0.5"
            y2="0.5" />
          <text
            dy="0.32em"
            fill="#000"
            x="-9"
            y="0.5">
            Label
          </text>
        </g>
        <g
          className="tick"
          transform="translate(0,360)">
          <line
            stroke="#000"
            x2="-6"
            y1="0.5"
            y2="0.5" />
          <text
            dy="0.32em"
            fill="#000"
            x="-9"
            y="0.5">
            Label
          </text>
        </g>
      </g>
      <g
        className="axis axis--x"
        fill="none"
        fontFamily="sans-serif"
        fontSize="10"
        textAnchor="middle"
        transform="translate(0,450)">
        <path
          className="domain"
          d="M0,6V0H890V6"
          stroke="#000"
          style={
            Object {
              "fill": "none",
              "shapeRendering": "crispEdges",
              "strokeWidth": 1,
            }
          } />
        <g
          className="tick"
          transform="translate(178,0)">
          <line
            stroke="#000"
            x1="0.5"
            x2="0.5"
            y2="6" />
          <text
            dy="0.71em"
            fill="#000"
            x="0.5"
            y="9">
            Label
          </text>
        </g>
        <g
          className="tick"
          transform="translate(356,0)">
          <line
            stroke="#000"
            x1="0.5"
            x2="0.5"
            y2="6" />
          <text
            dy="0.71em"
            fill="#000"
            x="0.5"
            y="9">
            Label
          </text>
        </g>
        <g
          className="tick"
          transform="translate(534,0)">
          <line
            stroke="#000"
            x1="0.5"
            x2="0.5"
            y2="6" />
          <text
            dy="0.71em"
            fill="#000"
            x="0.5"
            y="9">
            Label
          </text>
        </g>
        <g
          className="tick"
          transform="translate(712,0)">
          <line
            stroke="#000"
            x1="0.5"
            x2="0.5"
            y2="6" />
          <text
            dy="0.71em"
            fill="#000"
            x="0.5"
            y="9">
            Label
          </text>
        </g>
      </g>
    </g>
  </svg>
</div>
>>>>>>> 8cfaac105cba40e95ea519eaf4220f757ef8acf7
`;
